core/src/main/java/org/infinispan/persistence/factory/CacheStoreFactory.java

```diff
@@ -0,0 +1,24 @@
+package org.infinispan.persistence.factory;
+
+import org.infinispan.configuration.cache.StoreConfiguration;
+
+/**
+ * Creates Cache Store instances.
+ *
+ * <i>Needs to be implemented when loading Cache Stores from custom locations (e.g. custom location on the disk).</i>
+ *
+ * @author Sebastian Laskawiec
+ * @since 7.2
+ */
+public interface CacheStoreFactory {
+
+   /**
+    * Returns new instance based on {@link org.infinispan.configuration.cache.StoreConfiguration}.
+    *
+    * @param storeConfiguration Configuration to be processed.
+    * @return Instance configured by the {@link org.infinispan.configuration.cache.StoreConfiguration}.
+    */
+   Object createInstance(StoreConfiguration storeConfiguration);
```

Comment:
Unfortunately we can't do it. 

Currently `PersistenceManagerImpl` checks if retrieved object is an instance of `CacheLoader` or/and `CacheWriter`. ([see this line for example](https://github.com/infinispan/infinispan/pull/3317/files#diff-d93c51f709416559fc199f2b1687e2f7R567)). So we can't assume that we obtain only writers here. We actually may obtain a custom class which implements both `CacheWriter` and `CacheLoader`. 

I believe changing to `<T> T createInstance(StoreConfiguration storeConfiguration);` is the only improvement we can get here.